### PR TITLE
Bug: contention may let go of the more requests for the number of cores

### LINKED ISSUE #117: [BUG] ThrottlingChecker.DoCheck pass is not expected
## Issue Description

Type: *bug report*

### Describe what happened

ThrottlingChecker.DoCheck qps is not expected

### Describe what you expected to happen

ThrottlingChecker.DoCheck as expected

### How to reproduce it (as minimally and precisely as possible)

```go
func TestThrottlingChecker_DoCheckQueueingParallel(t *testing.T) {
	for i := 100; i >= 0; i-- {
		tc := NewThrottlingChecker(1000)
		var qps float64 = 5

		gc := 1000
		w := &sync.WaitGroup{}
		w.Add(gc)
		wg := &sync.WaitGroup{}
		wg.Add(gc)

		var passCount, waitCount, blockCount int32 = 0, 0, 0
		for i := 0; i < gc; i++ {
			go func() {
				w.Done()
				// for concurrence
				w.Wait()
				res := tc.DoCheck(nil, 1, qps)
				if res.IsPass() {
					atomic.AddInt32(&passCount, 1)
				}
				if res.IsBlocked() {
					atomic.AddInt32(&blockCount, 1)
				}
				if res.Status() == base.ResultStatusShouldWait {
					atomic.AddInt32(&waitCount, 1)
				}
				wg.Done()
			}()
		}
		wg.Wait()
		assert.Equal(t, int32(1), passCount)
		assert.Equal(t, int32(gc), waitCount+blockCount+passCount)
		// Non-strict mode may not be strictly accurate, so here we tolerate a delta.
		assert.InEpsilon(t, qps, waitCount, 1)
	}
}
```

### PR DESCRIPTION
**Pull request by ljun20160606**
…e number of cores

Modify store with cas.

### Does this pull request fix one issue?

Fixes #117 

### Describe how you did it

as commit

### Describe how to verify it

run unittest TestThrottlingChecker_DoCheckQueueingParallel

### Special notes for reviews

### CODE REVIEW

**Comment by sczyh30** [comment]
@ljun20160606 Could you please resolve conflicts and update according to the latest code in master branch?

**Comment by ljun20160606** [comment]
> @ljun20160606 Could you please resolve conflicts and update according to the latest code in master branch?

Done.

File: core/flow/tc_throttling.go [hunk 1]
```diff
@@ -65,21 +65,22 @@ func (c *ThrottlingChecker) DoCheck(_ base.StatNode, batchCount uint32, threshol
...
+	loadedLastPassedTime := atomic.LoadInt64(&c.lastPassedTime)
 	// Expected pass time of this request.
-	expectedTime := atomic.LoadInt64(&c.lastPassedTime) + intervalNs
+	expectedTime := loadedLastPassedTime + intervalNs
 	if expectedTime <= curNano {
-		// Contention may exist here, but it's okay.
-		atomic.StoreInt64(&c.lastPassedTime, curNano)
-		return nil
+		if swapped := atomic.CompareAndSwapInt64(&c.lastPassedTime, loadedLastPassedTime, curNano); swapped {
+			return base.NewTokenResultPass()
```
Comment (luckyxiaoqiang) [inline core/flow/tc_throttling.go:73]: Maybe return nil is better, so we can avoid memory allocation.

File: core/flow/tc_throttling_test.go [hunk 2]
```diff
@@ -76,16 +76,17 @@ func TestThrottlingChecker_DoCheckQueueingParallel(t *testing.T) {
...
 	wg := &sync.WaitGroup{}
 	gc := 24
 	wg.Add(gc)
 
-	var waitCount, blockCount int32 = 0, 0
+	var passCount, waitCount, blockCount int32 = 0, 0, 0
 	for i := 0; i < gc; i++ {
 		go func() {
 			res := tc.DoCheck(nil, 1, threshold)
+			if res.IsPass() {
```
Comment (luckyxiaoqiang) [inline core/flow/tc_throttling_test.go:87]: Add a latch before doing check may be better, which can let the goroutine concurrency more reasonable.

**Comment by luckyxiaoqiang** [comment]
`TestThrottlingChecker_DoCheckQueueingParallel` is used to test block and queueing in concurrency. Better a new unit test to test your change.

File: core/flow/tc_throttling_test.go [hunk 3]
```diff
@@ -100,3 +100,35 @@ func TestThrottlingChecker_DoCheckQueueingParallel(t *testing.T) {
...
+func TestThrottlingChecker_DoCheckParallelPass(t *testing.T) {
+	intervalMs := 10000
+	threshold := 50.0
+	timeoutMs := 0
+
+	tc := NewThrottlingChecker(nil, uint32(timeoutMs), uint32(intervalMs))
+
+	wg := &sync.WaitGroup{}
+	beginWg := &sync.WaitGroup{}
+	gc := 100000
```
Comment (sanxun0325) [inline core/flow/tc_throttling_test.go:113]: A large number of goroutine are responsible for the CI error.
```
=== RUN   TestThrottlingChecker_DoCheckParallelPass
race: limit on 8128 simultaneously alive goroutines is exceeded, dying
```

Comment (luckyxiaoqiang) [reply core/flow/tc_throttling_test.go]: I changed it to 512.

**Review by luckyxiaoqiang — APPROVED** [review]
LGTM.

**Review by sanxun0325 — APPROVED** [review]
LGTM

**Review by louyuting — APPROVED** [review]
LGTM

**Pull request merged**